Resolves #535: Improve the UX of modals on smaller screens

pieer opened the pull request:
Make all modals on smaller screens (including tablets) fullscreen.
Increase the size of the close button on all screens, so it's easier to click/tap on it;
Decrease the size of headings in smaller screens, to optimize space;
Adjust spacings;

+ add design system example for all the configuration
+ Remove Bulma modal as per new directive

pieer committed:
Update menu style

pieer committed:
Add transition

pieer committed:
Polish transitions

pieer committed:
Fix linting

pieer committed:
Fix test by adding temporary then inbox and a logout button in the profile

pieer committed:
Add themes colors

pieer committed:
Add comments for deep vue css selector

pieer committed:
Lint code

pieer committed:
New design part 1

pieer committed:
Fix merge

pieer committed:
Redesign modal

pieer committed:
Add transition

pieer committed:
Lint code

mmbotelho commented:
Hello Pierre!

First off, I really like the animation you added to the close button on hover, great job on that. Everything is looking great, except for a few things that I've noticed so far:

- The login and register modals should have the main content be wider (max 400px). [I made an example here ](https://www.figma.com/file/ZR0DRPhhUuHdJtf7PbGAM6tW/Group-Income?node-id=1918%3A29);
- The "escape" button should close the modals;
- The footer doesn't need to be sticky - It only needs to be positioned absolutely on the bottom of the modal card (on desktop - in smaller screens it should stay below the content like it normally would). This means that the footer should never overlap the main content of the modal. The cases that were designed so far only include text and an icon, never buttons or any other elements. See the "[simple modal with subtitle and footer](https://www.figma.com/file/ZR0DRPhhUuHdJtf7PbGAM6tW/Group-Income?node-id=1329%3A0)" here;
- In the "tablet" version, the modal should take up the entire screen.The first image is how it is currently implemented, the second one is how it should look like:
<img width="300px" alt="Captura de ecrã 2019-03-17, às 16 12 39" src="https://user-images.githubusercontent.com/22004566/54493618-581ae380-48d2-11e9-8020-b5cdf925ae94.png">
<img width="300px" alt="Captura de ecrã 2019-03-17, às 16 19 02" src="https://user-images.githubusercontent.com/22004566/54493623-6bc64a00-48d2-11e9-8c72-6ed4b41dfcbf.png">

- When I use the browser back button, the page behind the modal changes, but the modal remains open on top of it;
- Is there a way for us to change the url when a modal is open, so that it is linkable? I think this can also solve the browser back button problem, but I'm not sure.

Let me know if you have any questions!

pieer committed:
Fix modal design on tablet

pieer committed:
Update URL according to modal state

pieer committed:
Sticky footer not sticky on mobile

pieer committed:
Lint code

pieer committed:
Fix keyboards events

pieer commented:
Thanks for your feedback Margarida:
I've updated the branch according to your comments.

pieer committed:
Remove keyboard listener on destroy event

mmbotelho commented:
Hello @pieer ! Here's some other things I noticed:

- The different breakpoints are perfect now, thank you! 🎉 
- The login modal title should say "Login" instead of "Log In"
- The "Income Details" modal does not have the correct styling. Are you leaving it like this on purpose? We will be changing it to a fullscreen modal so it's not really a problem;
- Our buttons and links don't seem to have a style for the focused state. Should I open a new issue and design a style for this?
- The **app/design-system** page still has a sticky footer example. I see that you included buttons and errors on the footer. I am building things slightly differently on Figma - for me buttons and errors are a part of the main content and the footer only has an informative label. Maybe I've been calling it footer and that's not the correct name - let me know what you think about this. Also, the styling of the footer on smaller screens is not according to mockups yet.
- There are some styling changes I would like to make to the signup and login pages, but I will create a new issue since it's not directly related to the modals themselves.

pieer committed:
Replace Log in to login

pieer committed:
Fix footer modal style on mobile

pieer committed:
Move button and error to body

pieer committed:
Remove sticky footer

pieer commented:
@mmbotelho I've change the "Log In" to "Login"
I've removed the option for the sticky footer
I've moved button and errors in the body

For the rest, it look like this will be part of other tickets

mmbotelho commented:
Perfect @pieer! I have nothing else to add.

Thank you 🎉

taoeffect commented on frontend/views/components/Modal/ModalTemplate.vue:
```diff
@@ -49,27 +50,240 @@ export default {
-      if (isEscapeKey || isCloseButtonClick) {
-        sbp('okTurtles.events/emit', 'close-modal')
-      }
+      sbp('okTurtles.events/emit', 'close-modal')
```
This needs to be `CLOSE_MODAL`, not `'close-modal'`

taoeffect commented on frontend/views/components/Modal/ModalTemplate.vue:
```diff
@@ -49,27 +50,240 @@ export default {
     closeModal () {
       this.isActive = false
+      if (location.href.indexOf('?modal') > 0) {
+        history.pushState(null, null, location.href.split('?modal=')[0])
```
This is not the best or simplest or bug-free way to do this. The issue here is that by splitting the string at `?modal=`, and then passing in the first element of the array, the code fails to:

- Account for other URL variables that might come after it
- It assumes that `modal` will always be the first part of the URL, but the code on line 32 of Modal.vue simply appends `?modal=` to the URL, which could result in gibberish invalid URLs when the URL already has arguments appended  to it

So this approach is not workable. Either you must use (or create) a valid URL manipulating library, or a non-URL approach must be taken.

Another issue is the use of `window.localStorage` to keep track of the state of the popup. That is code smell. We have two kinds of state management in the app, and they are both described in our [Style Guide Data Model Rules](https://github.com/okTurtles/group-income-simple/blob/master/docs/Style-Guide.md#group-income-data-model-rules). Adding a third type of state management haphazardly here is again, code smell.

I recommend ignoring the back button for now. We might not use it anyway. And if we do, we will have to design and implement a coherent system for it, which will be its own issue and its own PR.

pieer replied:
Good comments Greg!
You raised a fair point about the ?modal=.
I did that intentionally however because I always want this to be the first argument, no matter what. I don't see any reason why other arguments should be coming before, and if there is it will raise that there is a problem in the URL structure. My recommendation is that we keep it this way :) Otherwise the simple solution to this problem is to remove the '?'. I don't see atm what a 'manipulating library' would do. Maybe you can elaborate on that point?

Url approach is required by @mmbotelho in the ticket for a valid point. I'm personally very annoyed when devs don't let me use the back option on mobile phone to remove the popup (which appear as a new page on mobile).
Also, we might want to redirect people to page with modals: for example if we link to a particular section of the settings in the user documentation.

An alternative is to NOT use modal, which is from my point of view the best approach, but that mean changing quite a bite the UX and design that @mmbotelho did.

Concerning the localStorage I agree. I should use the store. That was an old programmer reflex  when dealing with browser history... ;)

I don't think we need another ticket for this as this is a very quick change.

taoeffect replied:
Modals should not have URLs unless they are absolutely needed. Remember, this is an app, not a normal website. Modals do not have URLs in Slack, Figma, Discord, or any other app that I'm aware of, and they probably won't have URLs in Group Income either.

However, *if* we do add URLs for modals, then we will do it very carefully, and so an issue will need to be opened for that, and a clear problem description written up.

In general, using the URL to pass parameters is something that will require a separate issue and separate PR for, because it has to be well thought out. Simply manipulating the URL randomly with string concatenation in random places in the code is a definite no-go, because it will lead to ugly code, and lots and lots of bugs.

taoeffect replied:
>  I don't see atm what a 'manipulating library' would do. Maybe you can elaborate on that point?

A URL manipulating library makes it possible to manipulate the URL (exactly what you're trying to do), but safely, and without directly touching the URL and manipulating it yourself. It ensures the URL is well-formed, and makes it easy to add/remove/get parameters and their values.

However, again, we should not dive into using such a thing without a *lot* of thought and discussion first.

pieer replied:
Ok! That was only my recommendations... I'll remove my solution and I'll let you create a ticket for that with all the requirements.

taoeffect replied:
Yet another critical reason why a lot of thought must go into this (and a separate issue created, if they are deemed necessary), is that manipulating the URL and browser history manually as  this PR does **will** introduce severe conflicts with any other existing URL and history manipulating libraries we use.

We already use one such library, and this PR would be in conflict with it: [`vue-router`](https://router.vuejs.org)

VueRouter has its own way of passing parameters through the URL and interacting with them. We would need to heavily think about and weigh the pros & cons of different approaches, i.e. whether this use case would need URL parameters at all. If it was deemed necessary, then we would need to decide how we would use them, i.e. via VueRouter's API, or via a different library. If via a different library, we would have to research how that library should interact with VueRouter and VueX, etc.

taoeffect replied:
>  I'll remove my solution and I'll let you create a ticket for that with all the requirements.

👍 

Re the ticket: I'm not convinced that the Modal needs to be tied to the URLs in the first place, so I wouldn't know what kind of ticket to create. So if you have a Problem/Solution statement, or if @mmbotelho does, then one of you should create that ticket and we can discuss it

pieer replied:
I've fixed all the history manipulation using basic vue router watch and query parameter.
Great work on reviewing the code Greg!

taoeffect reviewed (changes requested):
Changes needed:

- No `window.localStorage.setItem('modal', true)`
- No `window.addEventListener('popstate' ...`

See comment on line 74 of ModalTemplate for details.

EDIT: I'm still reviewing the rest and will post more comment(s).

taoeffect commented on frontend/views/components/Modal/Modal.vue:
```diff
@@ -15,15 +15,32 @@ export default {
+          window.localStorage.setItem('modal', false)
+          sbp('okTurtles.events/emit', CLOSE_MODAL)
+        }
+        window.removeEventListener('popstate', null, null)
```
See [comment below](https://github.com/okTurtles/group-income-simple/pull/543/files?file-filters%5B%5D=.js&file-filters%5B%5D=.vue#r267117137)

pieer replied:
Good point I'll use the store.

taoeffect commented:
OK, other than the issue above, and one more possible issue that I'll mention here, this PR looks good, you did a **great job on implementing the CSS/UI of these new Modals!** 😄 👍 

I don't know if this is something the other PRs address, but on Firefox Developer Edition I found that clicking "Propose Change" on the Mincome brings up weird styling/spacing in the modal:

![Screen Shot 2019-03-19 at 6 25 34 PM](https://user-images.githubusercontent.com/138706/54646174-ac5dc900-4a74-11e9-9f91-54c43ea43010.jpg)

mmbotelho commented:
@dotmacro just confirmed on Figma that "Log in" is the correct term instead of "Login", sorry about the confusion! The comment she left on Figma was the following:

> Should be "log in" (as in "please log in" or "please sign in"). Without the space, it's a noun, like "credentials" or "username". 

So both the title of the modal and the button should say "Log in". I am sorry for the confusion there!

Regarding the modal URL thread, I still think it's a usability plus if we have a different URL. [User testing has proved that users perceive modals as a new page and expect the back button to lead them to what they were seeing before](https://baymard.com/blog/back-button-expectations). Not having a URL also makes sharing a specific view more difficult for the user (e.g.: the user wants to let someone know where they can change their income information).
Of course this is not a critical issue, it's more of an improvement that we can make later, if we all agree that this is important. If it has more cons than pros, from a technical perspective, I'm ok with moving on with how it is right now 😄

taoeffect commented:
> Regarding the modal URL thread, I still think it's a usability plus if we have a different URL. User testing has proved that users perceive modals as a new page and expect the back button to lead them to what they were seeing before. Not having a URL also makes sharing a specific view more difficult for the user (e.g.: the user wants to let someone know where they can change their income information).
Of course this is not a critical issue, it's more of an improvement that we can make later, if we all agree that this is important. If it has more cons than pros, from a technical perspective, I'm ok with moving on with how it is right now 😄

@mmbotelho yeah, this functionality is important in a general case and is being tracked as `Level:Advanced` issue #509. I've added the use cases of linking to specific settings within a modal and back/forward button support to that issue as well.

However, since there are a lot of ways of implementing that, and only a very small number of ways to do it right (without breaking things or making the codebase unmanageable), it will have to be done later as part of its own PR request, and only after a lot of research into the proper way to integrate it into the app without causing conflicts with vue-router, vuex, any state machines we add, and anything else.

mmbotelho commented:
Makes sense @taoeffect, thank you for clarifying!

pieer committed:
Replace local storage by the store

pieer committed:
Replace history manipulation and use router query instead

pieer committed:
Replace 'close-modal' by CLOSE_MODAL

pieer committed:
Fix buttons positions

pieer committed:
Fix initial modal loading

pieer committed:
Remove unecessary modal storage

taoeffect commented:
@pieer These latest changes are a significant improvement! Fantastic job! Merging! 🎉 👍

taoeffect reviewed (approved):
@pieer These latest changes are a significant improvement! Fantastic job! Merging! 🎉 👍
